docs: document workload resource mapping (#1139)
redhat-developer/service-binding-operator: docs/userguide/modules/creating-service-bindings/pages/custom-path-injection.adoc

spec:
    secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581
....

== Workload resource mapping

> The specification also provides a way of defining exactly where binding data needs to be projected

Author: Fixed.

into, which may work in situations where the above methods won't work. This is done by defining
`ClusterWorkloadResourceMapping.servicebindings.io` resources, which define where binding data gets

> projected for a workload type.  This works for both CoreOS and specification service binding

Reviewer: ```suggestion
projected for a given workload kind.  This works for both CoreOS and specification service binding
```

Author: Fixed.

> resources.

Author: Fixed.

> The schema of these resources looks like the following:

Author: Fixed.

[source,yaml]

> ....

Author: Fixed.

apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
  name: # string <1>

>   generation: # int64, defined by the Kubernetes control plane

Reviewer: Is this `metadata.generation` something that a user needs to include or set for the resource (YAML/JSON) or is managed/generated by Kubernetes/Openshift itself automatically (similarly as `creationTimestamp`)?

In case it is generated, I would not mention it here.

Author: It's autogenerated, and the user doesn't have to specify it.  I had it in because the spec mentions it; I've taken it out for now.

  ...
spec:
  versions: # required, at least one entry must be defined
  - version: # string <2>
    annotations: # string (Fixed JSONPath), optional <3>
    containers: # optional
    - path: # string (JSONPath) <4>
      name: # string (Fixed JSONPath), optional <5>
      env: # string (Fixed JSONPath), optional <6>
      volumeMounts: # string (Fixed JSONPath), optional <7>
    volumes: # string (Fixed JSONPath), optional <8>

> ....

Author: Fixed.

<1> the name of this resource (must be qualified as `plural.group`, i.e. `cronjobs.batch` for

> CronJobs)

Author: Fixed.

> <2> the version of the resource we're mapping (can match all unspecified versions with the wildcard "*")

Author: Fixed.

<3> where annotations live with a fixed JSONPath (default: `.spec.template.spec.annotations`)
<4> where containers live with a JSONPath (if no entries are defined, SBO defaults to two paths:
`.spec.template.spec.containers` and `.spec.template.spec.initContainers` with all other fields

> set as their default).

Author: Fixed.

<5> corresponds to the `.name` field in a pod. Defaults to `.name`.

> <6> corresponds to the `.env` environment variables. Defaults to `.env`.

Reviewer 2: ```suggestion
<6> Optional: Identifier of the `.env` environment variables. The default value is `.env`.
```

Author: Fixed.

> <7> corresponds to where the volume mounts of a pod lives. Defaults to `.volumeMounts`.

Reviewer 2: ```suggestion
<7> Optional: Custom location of the volume mounts of a pod. The default value is `.volumeMounts`.
```

Author: Fixed.

> <8> where volumes live with a fixed JSONPath (default: `.spec.template.spec.volumes`)

Reviewer 2: ```suggestion
<8> Optional: Custom location of the volumes with a fixed JSONPath. The default value is `.spec.template.spec.volumes`.
```

Author: Fixed.

> In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only two
> operations: field lookup (i.e. `.spec.template`) and indexing (i.e. `.spec['template']`).  All other
> operations are forbidden.

Reviewer 2: ```suggestion
[IMPORTANT]
====
In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only the following operations: 

. Field lookup, for example. `.spec.template` 
. Indexing, for example, `.spec['template']`

All other operations are forbidden.
====
```

Author: Fixed.

> In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only two
> operations: field lookup (i.e. `.spec.template`) and indexing (i.e. `.spec['template']`).  All other
> operations are forbidden.

Reviewer: ```suggestion
In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only two
operations: field lookup (i.e. `.spec.template`) and array indexing (i.e. `.spec['template']`).  All other
operations are forbidden.
```

Author: Fixed.

> Most of these fields are optional, and SBO assumes PodSpec-compatible defaults when not specified.
> SBO requires that each of these fields is semantically equivalent to the corresponding field in a Pod
> deployment; failing to maintain this may result in unexpected behavior.

Reviewer 2: ```suggestion
Most of these fields are optional. When not specified, the {servicebinding-title} assumes defaults that are compliant with the `PodSpec workload resources.
+ 
[NOTE]
====
Each of these fields must be semantically equivalent to the corresponding field in a pod deployment.
deployment. Otherwise, this might result in unexpected behavior.
```
Question: 

- Can you state an example of what such unexpected behavior might look like for general understanding purposes?
- Unexpected behavior of what? Please clarify.

Author: If a workload resource is not semantically compatible with a pod template, we can't make any guarantees about the behavior a user observes.  This statement is meant to be a warning along those lines.

Author: On consideration, I've decided to reword this warning more toward the need for structural equivalence.  I think talking about semantic equivalence is less important structural compatibility.  If structural compatibility is not maintained, SBO will throw an error when binding is attempted, and users of SBO should be aware of this limitation.

> Using this schema, we can write an example mapping for `CronJob.batch/v1` resources:

Reviewer 2: ```suggestion
Using this schema, you can write an example mapping for the `CronJob.batch/v1` resources:

.Example: 
```

Author: See Pavel's comments on this.

Author: Fixed.

> Using this schema, we can write an example mapping for `CronJob.batch/v1` resources:

Reviewer: ```suggestion
Using the following YAML, we can define an example mapping for `CronJob.batch/v1` resources:
```

Author: ```suggestion
Using the above schema, we can define the following YAML as an example mapping for `CronJob.batch/v1` resources:
```

Thoughts?

[source,yaml]

> ....

Reviewer 2: ```suggestion
----
```

Author: On consideration, I don't understand why this change is necessary.  The other yaml listings in this file use `....` as delimiters, so why do I need to use `----` here?

Reviewer 2: This is the asciidoc convention we follow in downstream docs. As we take some good amount of content from upstream for downstream purposes, it is good to practice the same thing here so both are in sync.

Author: Fixed for new instances.

apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
 name: cronjobs.batch
spec:
  versions:
  - version: "v1"
    annotations: .spec.jobTemplate.spec.template.metadata.annotations
    containers:
    - path: .spec.jobTemplate.spec.template.spec.containers[*]
    - path: .spec.jobTemplate.spec.template.spec.initContainers[*]
    volumes: .spec.jobTemplate.spec.template.spec.volumes

> ....

Reviewer 2: ```suggestion
----
```

Author: Fixed.

NOTE: There is some behavior in the CoreOS API that doesn't exist in the specification API:

> . If a service binding with `bindAsFiles: false` is created with one of these mappings, then

Reviewer: ```suggestion
. If a `ServiceBinding` resource with `bindAsFiles: false` is created together with one of these mappings, then
```

Author: Fixed.

> NOTE: There is some behavior in the CoreOS API that doesn't exist in the specification API:
> 
> . If a service binding with `bindAsFiles: false` is created with one of these mappings, then
> environment variables will be projected into `.envFrom` underneath each `path` field specified in the
> corresponding `ClusterWorkloadResourceMapping` resource.
> . Both this and the `containersPath` field documented above can be used at the same time.  This would
> be equivalent to adding a container entry with `path: $(containersPath)`.

Reviewer 2: ```suggestion
[NOTE]
====
The following behavior occurs only in the `binding.operators.coreos.com` API group:

* If a service binding with the `bindAsFiles: false` flag value is created with one of these mappings, then environment variables are projected into the `.envFrom` field under each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
* Both the fields `.envFrom` and `containersPath` can be in use at the same time. This is equivalent to adding a container entry with the `path: $(containersPath)` attribute.
```
Question:
- In the sentence at L207: "Both this ..", What do you mean exactly by "this"? Please clarify and provide inputs.
- Again, in "...can be used at the same time", What do you mean exactly? Used for what, who is using? Please clarify and provide inputs.
- The 2nd bullet item is very confusing and difficult to comprehend. Please rewrite for clarity.

Author: > In the sentence at L207: "Both this ..", What do you mean exactly by "this"? Please clarify and provide inputs.

By "this", I mean `ClusterWorkloadResourceMapping` resources.

> Again, in "...can be used at the same time", What do you mean exactly? Used for what, who is using? Please clarify and provide inputs.

It is possible for a cluster administrator to specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding` resource.  SBO will attempt to project binding information into the locations specified in both of these methods.  That is the behavior I was attempting to document.

> The 2nd bullet item is very confusing and difficult to comprehend. Please rewrite for clarity.

I'll take another attempt at it - I agree it's not the clearest.

Author: I've taken another attempt at writing this bullet point, PTAL.

spec:
    secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581
....

== Workload resource mapping

[NOTE]
====
Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
====

> The specification also provides a way of defining exactly where binding data needs to be projected into.  This method may be used when the preceding methods would not be able to configure custom path locations correctly.  You may define `ClusterWorkloadResourceMapping.servicebindings.io` resources, which specifies where binding data will be projected for a given workload kind.

Reviewer 2: @sadlerap Any specific reason why the suggested changes are not applied here? I do not also see a comment from you as you had mentioned highlighting the changes you needed to make.

.Example: Schema of the `ClusterWorkloadResourceMapping` resource
[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
  name: # string <1>
  ...
spec:
  versions: # required, at least one entry must be defined
  - version: # string <2>
    annotations: # string (Fixed JSONPath), optional <3>
    containers: # optional
    - path: # string (JSONPath) <4>
      name: # string (Fixed JSONPath), optional <5>
      env: # string (Fixed JSONPath), optional <6>
      volumeMounts: # string (Fixed JSONPath), optional <7>
    volumes: # string (Fixed JSONPath), optional <8>
----
<1> Name of the `ClusterWorkloadResourceMapping` resource, which must be qualified as the `plural.group` of the mapped workload resource.  For example, `cronjobs.batch` for CronJobs.
<2> Version of the resource that is being mapped.  Any version that is not specified can be matched with the "*" wildcard.
<3> Optional: Identifier of the `.annotations` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.annotations`.

> <4> Identifies where containers live, specified with a JSONPath.  If no entries are defined, the {servicebinding-title} defaults to two paths: `.spec.template.spec.containers` and `.spec.template.spec.initContainers`, with all other fields set as their default.

Reviewer 2: @sadlerap Any specific reason why the suggested changes are not applied here? I do not also see a comment from you as you had mentioned highlighting the changes you needed to make.

<5> Optional: Identifier of the `.name` field in a container, specified with a fixed JSONPath.  The default value is `.name`.
<6> Optional: Identifier of the `.env` field in a container, specified with a fixed JSONPath.  The default value is `.env`.
<7> Optional: Identifier of the `.volumeMounts` field in a container, specified with a fixed JSONPath.  The default value is `.volumeMounts`.
<8> Optional: Identifier of the `.volumes` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.volumes`.

[IMPORTANT]
====
In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only the following operations:

. Field lookup: `.spec.template`
. Array indexing: `.spec['template']`

All other operations are forbidden.
====

[NOTE]
====

> * Most of these fields are optional, and the {servicebinding-title} assumes PodSpec-compatible defaults when they are not specified.

Reviewer 2: ```suggestion
* Most of these fields are optional, and when not specified, the {servicebinding-title} assumes defaults that are compliant with the `PodSpec workload resources.
```
Suggest writing sentences that give complete details. Avoid ambiguous pronouns - `they`.

Author: I thought the usage of `they` in this context was clear.  Would this be better?
```suggestion
* Most of these fields are optional.  When they are not specified, the {servicebinding-title} assumes defaults compatible with `PodSpec` compatible workload resources.
```

> * Most of these fields are optional.  When they are not specified, the {servicebinding-title} assumes defaults compatible with `Pod` resources.

Reviewer 2: ```suggestion
* Most of these fields are optional. When they are not specified, the {servicebinding-title} assumes defaults compatible with `Pod` resources.
```

Reviewer 2: @sadlerap Thanks for the change. Just remove the extra space.

Reviewer 2: @sadlerap Are these `Pod` resources or `PodSpec` resources? Please clarify.

> * The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a Pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding information into such a workload may result in unexpected behavior from the {servicebinding-title}.

Reviewer 2: ```suggestion
* The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding data into such a workload might result in unexpected behavior from the {servicebinding-title}.
```

Author: Fixed.

====

Using the previous schema, you can define the following YAML as an example mapping for `CronJob.batch/v1`
resources:

[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
 name: cronjobs.batch
spec:
  versions:
  - version: "v1"
    annotations: .spec.jobTemplate.spec.template.metadata.annotations
    containers:
    - path: .spec.jobTemplate.spec.template.spec.containers[*]
    - path: .spec.jobTemplate.spec.template.spec.initContainers[*]
    volumes: .spec.jobTemplate.spec.template.spec.volumes
----

> [NOTE]
> ====
> There is some behavior in the CoreOS API that does not exist in the specification API:
> 
> . If a `ServiceBinding` resource with `bindAsFiles: false` is created with one of these mappings, then environment variables will be projected into `.envFrom` underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
> . Both `ClusterWorkloadResourceMapping.servicebinding.io` resources and the `.spec.application.bindingPath.containersPath` field will be considered for binding purposes.  This behavior would be equivalent to adding an entry to the corresponding `ClusterWorkloadResourceMapping` resource with `path: $containersPath`, with all other values taking their default value.
> ====

Reviewer 2: ```suggestion
[NOTE]
====
The following behavior occurs only in the `binding.operators.coreos.com` API group:
. If a `ServiceBinding` resource with the `bindAsFiles: false` flag value is created together with one of these mappings, then environment variables are projected into the `.envFrom` field underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
. As a cluster administrator, you can specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding.bindings.coreos.com` resource for binding purposes. The {servicebinding-title} attempts to project binding data into the locations specified in both of these methods. This behavior is equivalent to adding a container entry to the corresponding `ClusterWorkloadResourceMapping` resource with the `path: {containersPath}` attribute, with all other values taking their default value.
====
```
The Verbatim is marked as fixed and resolved, but I am not seeing the suggested changes. Hence suggesting once again. It was already there in the [comment given previously](https://github.com/redhat-developer/service-binding-operator/pull/1139#discussion_r876104617). Ensure to apply the changes suggested as is, as the review comments are given from the language guidelines perspective that must adhere to RH standards.

Author: I can't use `path: {containersPath}`, since `asciidoctor` correctly complains that the `{containersPath}` reference is not defined.
```suggestion
[NOTE]
====
There is some behavior in the CoreOS API that does not exist in the specification API:

. If a `ServiceBinding` resource with `bindAsFiles: false` is created with one of these mappings, then environment variables will be projected into `.envFrom` underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
. As a cluster administrator, you can specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding.bindings.coreos.com` resource for binding purposes.  The {servicebinding-title} attempts to project binding data into the locations specified in both of these methods.  This behavior is equivalent to adding a container entry to the corresponding `ClusterWorkloadResourceMapping` resource with the `path: $containersPath` attribute, with all other values taking their default value.
====
```

spec:
    secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581
....

== Workload resource mapping

[NOTE]
====

> . Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
> . You must namespace `ClusterWorkloadResourceMapping` resources under the `servicebinding.io` API group.  However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

Reviewer 2: ```suggestion
* Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
* You must namespace `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group.  However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.
```
Suggestion: Bulleted list, please.

Author: Fixed.

<3> Optional: Identifier of the `.annotations` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.annotations`.
<4> Identifier of the `.containers` and `.initContainers` fields in a pod, specified with a JSONPath.  If no entries under the `containers` field are defined, the {servicebinding-title} defaults to two paths: `.spec.template.spec.containers[*]` and `.spec.template.spec.initContainers[*]`, with all other fields set as their default.  However, if you specify an entry, then you must define the `.path` field.
<5> Optional: Identifier of the `.name` field in a container, specified with a fixed JSONPath.  The default value is `.name`.
<6> Optional: Identifier of the `.env` field in a container, specified with a fixed JSONPath.  The default value is `.env`.
<7> Optional: Identifier of the `.volumeMounts` field in a container, specified with a fixed JSONPath.  The default value is `.volumeMounts`.
<8> Optional: Identifier of the `.volumes` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.volumes`.

> [IMPORTANT]

Reviewer 2: Question: Is this admonition only for L172?

Suggestion: If yes, then please put `+` in L173, so that the admonition is indented under the Example section.

Author: The `[IMPORTANT]` section is supposed to be for the entire list of callouts preceding it, not just for the last one.

====
In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only the following operations:

. Field lookup: `.spec.template`
. Array indexing: `.spec['template']`

All other operations are forbidden.
====

[NOTE]
====
* Most of these fields are optional.  When they are not specified, the {servicebinding-title} assumes defaults compatible with `PodSpec` resources.
* The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding data into such a workload might result in unexpected behavior from the {servicebinding-title}.
====

Using the previous schema, you can define the following YAML as an example mapping for `CronJob.batch/v1`
resources:

> 

Reviewer 2: Add after L192:

.Example: Mapping for the `CronJob.batch/v1` resources

Author: Fixed.

spec:
    secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581
....

== Workload resource mapping

NOTE: Workload resource mapping is available for the secondary workloads of the ServiceBinding custom

> resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.

Author: Shouldn't this be:
```asciidoc
[NOTE]
====
Workload resource mapping is available for the secondary workloads of the `ServiceBinding` custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
====
```

This is more consistent with the other documentation I have.

> resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.

Reviewer 2: Yes, you are correct!

> resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.

Reviewer 2: @sadlerap yes, you are correct!

> resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.

Author: Fixed.

The specification also provides a way of defining exactly where binding data needs to be projected
into, which may work in situations where the above methods won't work.  This is done by defining
`ClusterWorkloadResourceMapping.servicebindings.io` resources, which define where binding data gets
projected for a given workload kind.  This works for both CoreOS and specification service binding

> resources.

Author: Should be fixed now, PTAL.

> .Example Schema of the `ClusterWorkloadResourceMapping` resource

Author: Fixed.

[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
  name: # string <1>
  ...
spec:
  versions: # required, at least one entry must be defined
  - version: # string <2>
    annotations: # string (Fixed JSONPath), optional <3>
    containers: # optional
    - path: # string (JSONPath) <4>
      name: # string (Fixed JSONPath), optional <5>
      env: # string (Fixed JSONPath), optional <6>
      volumeMounts: # string (Fixed JSONPath), optional <7>
    volumes: # string (Fixed JSONPath), optional <8>
----

> <1> Name of this resource, which must be qualified as `plural.group`.  For example, `cronjobs.batch` for CronJobs.

Author: Fixed.

> <2> Version of the resource we're mapping.  Any version not specified can be matched with the "*" wildcard.

Author: Fixed.

> <3> Optional: where annotations live, specified with a fixed JSONPath.  The default value is `.spec.template.spec.annotations`.

Author: Fixed.

> <4> Optional: where containers live, specified with a JSONPath.  If no entries are defined, the {servicebinding-title} defaults to two paths: `.spec.template.spec.containers` and `.spec.template.spec.initContainers` with all other fields set as their default).

Author: Adjusted wording.  A few notes:
- This field is not optional, and I've adjusted this wording to reflect that.
- This field is not a fixed JSONPath; JSONPath is the correct wording here.

> <4> Optional: where containers live, specified with a JSONPath.  If no entries are defined, the {servicebinding-title} defaults to two paths: `.spec.template.spec.containers` and `.spec.template.spec.initContainers` with all other fields set as their default).

Reviewer 2: Left some suggestions, PTAL

<5> Optional: Identifier of the `.name` field in a pod.  The default value is `.name`.

> <6> Optional: Identifier of `.env` environment variables.  The default value is `.env`.

Author: Fixed.

> <7> Optional: Identifier of the volume mounts of a pod lives.  The default value is `.volumeMounts`.

Author: I've adjusted the wording here to be more consistent with the other optional fields.  PTAL.

> <8> Optional: where volumes live, specified with a fixed JSONPath.  The defaults value is `.spec.template.spec.volumes`.

Author: @Srivaralakshmi thoughts?

[IMPORTANT]
====
In this context, a "Fixed JSONPath" is a subset of the JSONPath grammar that accepts only the
following operations:

. Field lookup: `.spec.template`
. Array indexing: `.spec['template']`

All other operations are forbidden.
====

[NOTE]
====

> Most of these fields are optional, and SBO assumes PodSpec-compatible defaults when not specified.
> SBO requires that each of these fields is structurally equivalent to the corresponding field in a Pod
> deployment.  For example, the contents of the `.env` field in a workload resource must be able to
> accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise,
> projecting binding information into such a workload may result in unexpected behavior from SBO.

Reviewer 2: ```suggestion
* Most of these fields are optional, and when not specified, the {servicebinding-title} assumes defaults that are compliant with the `PodSpec workload resources.
* The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding data into such a workload might result in unexpected behavior from the {servicebinding-title}.
```

Author: Fixed.

====

> Using the above schema, one can define the following YAML as an example mapping for `CronJob.batch/v1`
> resources:

Reviewer 2: ```suggestion
Using the previous schema, you can define the following YAML as an example mapping for the `CronJob.batch/v1`
resources:
```
Added modifications to adhere to IBM and RH writing standards.

Author: Fixed.

[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
 name: cronjobs.batch
spec:
  versions:
  - version: "v1"
    annotations: .spec.jobTemplate.spec.template.metadata.annotations
    containers:
    - path: .spec.jobTemplate.spec.template.spec.containers[*]
    - path: .spec.jobTemplate.spec.template.spec.initContainers[*]
    volumes: .spec.jobTemplate.spec.template.spec.volumes
----

> NOTE: There is some behavior in the CoreOS API that does not exist in the specification API:
> 
> . If a `ServiceBinding` resource with `bindAsFiles: false` is created with one of these mappings,
> then environment variables will be projected into `.envFrom` underneath each `path` field specified
> in the corresponding `ClusterWorkloadResourceMapping` resource.
> . Both `ClusterWorkloadResourceMapping.servicebinding.io` resources and the
> `.spec.application.bindingPath.containersPath` field will be considered for binding purposes.  This
> behavior would be equivalent to adding an entry to the corresponding `ClusterWorkloadResourceMapping`
> resource with `path: {containersPath}`, with all other values taking their default value.

Reviewer 2: ```suggestion
NOTE
====
The following behavior occurs only in the `binding.operators.coreos.com` API group:

. If a `ServiceBinding` resource with the `bindAsFiles: false` flag value is created together with one of these mappings, then environment variables are projected into the `.envFrom` field underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
. As a cluster administrator, you can specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding.bindings.coreos.com` resource for binding purposes. The {servicebinding-title} attempts to project binding data into the locations specified in both of these methods. This behavior is equivalent to adding a container entry to the corresponding `ClusterWorkloadResourceMapping` resource with the `path: {containersPath}` attribute, with all other values taking their default value.
====
```

Author: Fixed.
